### ISSUE TITLE
Insert doesn't send pk column

### ISSUE THREAD

**Issue by Marka83** [linq2db/linq2db.EntityFrameworkCore#39]
Hi,

After migrating from EF to Linq2db.EntityFramework everything worked as expected,
but I wanted to migrate DML also to Linq2Db, so that I could eventually remove EF completely.

From some reason, insert doesn't work.
I have some simple table :

```
[Table("...", Schema = "...")]
public class NaseljenoMjestoEntity
	{
		[Key]
		public int? NaseljenoMjestoId { get; set; }

		[Required]
		public string Naziv { get; set; }

               ...
        }
```
I've used this code to try to insert:

```
using (var context = _dbContext.CreateLinqToDbConnection())
			{
			    var result= context.Insert(entity);
			}
```
but it doesn't work.
Query generated for insert contains all columns except NaseljenoMjestoId, so naturally, it throws Cannot insert the value NULL into column 'NaseljenoMjestoId'...

NaseljenoMjestoId  is Nullable<int>, but changing it to int doesn't change anything.

We have more than 1000 entities, so I was hoping to reuse EF annotations so that I wouldn't have to regenerate all the entities (most of them have some kind of custom modifications).
It's ease to replace Key to PrimaryKey, but ForeignKey attribute doesn't have 'OtherKey' part, so it would be tedious to try to manually replace all ForeignKeys to Association.

What am I doing wrong?
Is this use case event possible (to create DataConnection from DbContext and do insert, based on EF annotations)?

I'm using linq2db.entityframeworkcore 3.0.0

Also, what is the difference between  DataConnection and DataContext?
What are different use cases for both of them?

**Comment by sdanyliv** [comment]
Could you please post how you configure one entity. Possibly we have missed something.
Anyway it should work without limitations.

**Comment by Marka83** [comment]
For this entity, I only used annotations.
I stripped all navigated properties but it still doesn't work.
I've tried few other simpler entities and insert works as expected.

How can I trace what is happening?

**Comment by Marka83** [comment]
I've found out something interesting.
I examined entityType and it seems like this column has ValueGenerated.OnAdd.
I don't know how is that possible.
This column is not Identity in database and I also never used 
[DatabaseGenerated(DatabaseGeneratedOption.Identity)].

I'll investigate it further before wasting any of you time.

**Comment by sdanyliv** [comment]
Yes, that's how we detect Identity columns. Maybe I'm wrong but I'm not so good in EF stuff.

https://github.com/linq2db/linq2db.EntityFrameworkCore/blob/master/Source/LinqToDB.EntityFrameworkCore/EFCoreMetadataReader.cs#L116

**Comment by Marka83** [comment]
It turns out, it something stupid on EF part:
https://www.learnentityframeworkcore.com/conventions

"If a property is named ID or <entity name>ID (not case-sensitive), it will be configured as the primary key. Entity Framework Core will prefer ID over <entity name>ID in the event that a class contains both"

**"EF Core will specify that the primary key column values are generated automatically by the database"**
This is simply stupid. I don't even know how it work at all.
I'll have to modify conventions.

**Comment by sdanyliv** [comment]
Thanks for the link, good to know. I wonder how many people are stuck on this.

**Comment by sdanyliv** [comment]
Anyway, linq2db can process any annotations, you just have to define custom MetadaDataReader for mapping schema. Code link above shows that.

**Comment by Marka83** [comment]
That may be the best way to proceed. I'll implement IMetaDataReader and remove EF completely.
Thank you for your time.

**Comment by sdanyliv** [comment]
Looks like you've got a pain with EF. What is most annoying? Just for statistic :)

**Comment by Marka83** [comment]
I was using EF since version 1 (some 11 years ago) to version 6.
It mostly did its job. It wasn't perfect, but I managed to get around some of the shortcomings.
Two years ago we decided to migrate our ERP system to .Net Core with EF Core and Angular.

At that time, some of the basic functionalities weren't implemented (like server-side groupby).
In contrast to previous versions of EF, with EF Core they decided not to send all queries to database.
If the query is too complicated to generate query or if they decide  that it would be better to fetch all to client side and do processing on application server, they would do that.

It resulted in some serious performance problem.
For example, group by was done completely on application side, so you would fetch all the data and group by on application side (they did manage in 2.2 to generate Sql query, but complex queries would still be run on client side). Also, in many cases EF would generate many requests to database instead of one request.

We managed to keep performance in check by tuning queries where ever we could.
Also, they used external library for parsing ExpressionTree.

Since 3.0, they removed that external library in favor of their new implementation.
Problem is, many of the things that worked in 2.2 didn't work in 3.0.

Another problem - they finally came to sense that client processing of data is generally not a good idea, so every query should execute on server.
If they cannot create query they would throw exception. In that case, you would have to manually chop query into smaller peaces and do some manual processing.

That was the point I already had enough.

In parallel with that problem, I wanted to extend EF functionalities like update or delete entities 
based on where condition. 
This isn't possible in EF because first you have to materialize all entities in EF and then update or
delete every entity and call SaveChanges().

That was the time when I find out about Linq2Db.
It was exactly something that I wanted to create (not from scratch, but by extending EF).

Currently, I'm trying to migrate whole application to Linq2Db step by step.
First I was thinking to try to use query side and after 2 problems (which you did fix in very short time) everything worked fine.
Problem is, when you query with Linq2Db, those entities are not tracked by EF, which creates another type of problems.
For example, if you have entity with multiple navigation properties and you "include" same navigated entity through different paths, EF will throw exception if you try to saveChanges and will say that entity with same key is already tracked by EF.
In order to fix that I would have to traverse whole graph and integrate same entities, which is really too much complication for nothing.

This is why I decided to try to completely migrate to Linq2Db.

My only problem with Linq2Db is that there is much less documentation and examples then there is for EF.
Also, community is much smaller so you can't find many posts on stackoverflow for example.

On the architectural side, I do believe that you made much better decisions then it is case with EF.
At this time, I'll try to replace EF completelly and solve problem as they surface.

**Comment by sdanyliv** [comment]
I see. BTW, keep in mind that there is no change tracker in linq2db.

**Comment by Marka83** [comment]
I know, this is what I like about linq2db.
Why would I need change tracking?
I have webapi, I create context/connection, insert/update/delete entity and dispose of context.

Change tracking is the main reason why they can't implement update/delete with where clause.
https://github.com/dotnet/efcore/issues/795#issuecomment-577383483

**Comment by sdanyliv** [comment]
We do not need, but people are used to this :) DDD design dictates such pattern.

**Comment by Marka83** [comment]
I would always trade that for you functionality and simplicity.
Keep up with good work.

**Comment by sdanyliv** [comment]
Closing for now, will reopen if needed.
